ecosystem/typescript/sdk_v2/src/bcs/deserializer.ts

```diff
@@ -0,0 +1,192 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+/* eslint-disable no-bitwise */
+import { MAX_U32_NUMBER } from "./consts";
+import { Uint128, Uint16, Uint256, Uint32, Uint64, Uint8 } from "./types";
+
+// The class must implement a static deserialize method.
+interface Deserializable<T> {
+  deserialize(deserializer: Deserializer): T;
+}
+
+export class Deserializer {
+  private buffer: ArrayBuffer;
+
+  private offset: number;
+
+  constructor(data: Uint8Array) {
+    // copies data to prevent outside mutation of buffer.
+    this.buffer = new ArrayBuffer(data.length);
+    new Uint8Array(this.buffer).set(data, 0);
```

Comment:
you don't have to set this to something?

Reply:
no, apparently the `new Uint8Array(this.buffer)` serves as a sort of mutable reference to the data in the `ArrayBuffer` in the heap. So `.set` directly mutates the value in `this.buffer`, changing its data to `data` starting at the byte at index `0`